# Lab 4: Q-table based reinforcement learning



Solve [`FrozenLake8x8-v1`](https://www.gymlibrary.dev/environments/toy_text/frozen_lake/) using a Q-table.


1. Import Necessary Packages (e.g. `gym`, `numpy`):

In [67]:
import gym
import numpy as np
import random
import pygame


2. Instantiate the Environment and Agent

In [68]:
# Enviroment
# gym.make('FrozenLake-v1', render_mode='human')
env = gym.make('FrozenLake-v1',is_slippery=False)
#Taxi-v3


In [69]:
# Agent

class Agent:
    def __init__(self):
        self.total_episodes = 25000
        self.max_steps = 100
        self.epsilon = 0.1
        self.learn_rate = 0.2
        self.gamma = 0.85

        self.epsilon = 1.0
        self.max_epsilon = 1.0
        self.min_epsilon = 0.01
        self.decay_rate = 0.01

    def act(self, obs,q_table):
        return np.argmax(q_table[obs,:])


    def learn(self,q_table, state):
        for episode in range(self.total_episodes):
            env.reset()
            state = env.action_space.sample()
            step = 0
            return_total = 0
            done = False
            
            while not done:
                exploration_explotation_trade = random.uniform(0, 1)
                if exploration_explotation_trade > self.epsilon:
                    # returns index of the bigges value 
                    # q_table[state, :] the colon is all the actions of the state
                    action = np.argmax(q_table[state, :])
                else:
                    action = env.action_space.sample()
                
                observation, reward, done, _, _ = env.step(action)

                
                q_table[state,action] = q_table[state,action] + self.learn_rate *(reward+self.gamma*
                                                                            np.max(q_table[observation, :]) - q_table[state, action])
                
               
                state = observation
                return_total += reward
                if done:
                    observation = env.reset
     

        self.epsilon = self.min_epsilon + (self.max_epsilon - self.min_epsilon) *np.exp(-self.decay_rate*episode)


3. Set up the QTable:

In [70]:

action_size = env.action_space.n
state_size = env.observation_space.n

q_table = np.random.randn(state_size, action_size)

print("State size ", state_size)
print("Action size ", action_size)
print(env.action_space)


State size  16
Action size  4
Discrete(4)


4. The Q-Learning algorithm training

In [71]:
agent = Agent()
obs = env.reset()[0]
agent.learn(q_table,obs)

5. Evaluate how well your agent performs
* Render output of one episode
* Give an average episode return

In [72]:
from pathlib import Path


rewards = []
test_episodes = 10

env = gym.make('FrozenLake-v1', is_slippery=False, render_mode='human')


for i in range(test_episodes):
    obs = env.reset()[0]
    step = 0
    return_total = 0
    done = False
    total_rewards = 0
    while not done:
        action = agent.act(obs, q_table)
        observation, reward, done, _, _ = env.step(action)
        return_total += reward
        if(done):

            rewards.append(total_rewards)
            break
        obs = observation

    print(return_total/(i+1))
env.close()



1.0
0.5
0.3333333333333333
0.25
0.2
0.16666666666666666
0.14285714285714285
0.125
0.1111111111111111
0.1


Gives output: 

1.0
0.5
0.3333333333333333
0.25
0.2
0.16666666666666666
0.14285714285714285
0.125
0.1111111111111111
0.1


Gives the outputs 

6. (<i>Optional</i>) Adapt code for one of the continuous [Classical Control](https://www.gymlibrary.dev/environments/classic_control/) problems. Think/talk about how you could use our  `Model` class from last Thursday to decide actions.